llvm/lib/Target/RISCV/MCTargetDesc/RISCVAsmBackend.cpp

```diff
@@ -803,6 +803,41 @@ void RISCVAsmBackend::maybeAddVendorReloc(const MCFragment &F,
   Asm->getWriter().recordRelocation(F, VendorFixup, VendorTarget, VendorValue);
 }
 
+static bool fixupGetsRelaxRelocation(const MCFixupKind Kind) {
+  switch (Kind) {
```

Comment:
Why does any fixup need special treatment? check-llvm-mc still passes if fixupGetsRelaxRelocation returns false.

Reply:
if we don't have this, we'll end up with R_RISCV_RELAX on compressed instructions which are in-range but require a relocation (i.e. reference something close but the other side of something relaxable). I'll make a testcase for this tomorrow.

Reply:
I added more to the linker-relaxation-pr150071.s test which will fail if this either always returns `true` or `false`.

The problem is about the over-approximation of `RelaxCandidate = true` in the MCCodeEmitter. Now, many more relocations are marked as linkerrelaxable, which is good, but those might be pc-relative resolvable, like a jump over non-relaxable data. In these cases we want to allow the fixup itself to resolve if possible (these will only be left if they are in range, when xqcilb is enabled).

And then we want to make sure, if a fixup is not resolved, that only fixups expecting R_RISCV_RELAX get that relocation, even if the fixup is marked linker relaxable.